Documentation/scalability/scalability.rst

```diff
@@ -0,0 +1,397 @@
+
+Since Kubernetes requires an ``etcd`` cluster, we have deployed v3.4.9.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+2. Deploy Prometheus, Grafana and Cilium
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We have used prometheus v2.18.1 and grafana v7.0.1 to retrieve and analyze
+``etcd``, ``kube-apiserver``, ``cilium`` and ``cilium-operator`` metrics;
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+3. Provision 2 workers nodes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This helped us to understand if our testing cluster was correctly provisioned
+and all metrics were being gathered.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+4. Deploy 5 namespaces with 25 deployments on each namespace
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* Each deployment had 1 replica each (125 pods in total).
+
+* To measure the resources **only** consumed by Cilium, all deployments used
+  the same base image ``k8s.gcr.io/pause:3.2``.
+
+* Since we provision a small number of pods in our small cluster we are able to
+  understand the CPU usage of Cilium when a small number of pods is provisioned:
+
+.. figure:: images/image_4_01.png
+
+The mark shows when the creation of 125 pods have started.
+As expected, we can see a slight increase of the CPU usage on both
+cilium-agents running and in the cilium operator. The Cilium agents
+peaked at 6.8% CPU usage on a 2vCPU machine.
+
+.. figure:: images/image_4_02.png
+
+For the memory usage we have not seen a significant memory growth in the
+Cilium agent. On the BPF memory side we do see it increasing due the
+initialization of some BPF maps for the pods that we have started.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+5. Provision 998 new nodes (total 1000 nodes)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. figure:: images/image_5_01.png
+
+The first mark represents the action of creating nodes, the second mark
+represents when 1000 Cilium pods were in ready state. It is expected that the
+CPU usage of Cilium increases since each Cilium agent receives events from
+Kubernetes whenever a new node is provisioned in the cluster. Once all nodes
+were deployed the average CPU usage was on average 0.15% on a 2vCPU node.
+
+.. figure:: images/image_5_02.png
+
+As we have increased the number of nodes in the cluster to 1000 it is expected
+to see a small growth of the memory usage in all metrics. However, it is
+relevant to point out that **the number of nodes does not have a significant
+growth in Cilium’s memory consumption in both control and dataplane.**
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+6. Deploy 25 more deployments on each namespace
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This will now bring us a total of
+``5 namespaces * (25 old deployments + 25 new deployments)=250`` deployments in
+the entire cluster.
+We did not install 250 deployments from the start since we only had 2 nodes and
+that would create 125 pods on each worker node and accordingly with the
```

Comment:
maybe:


We did not install 250 deployments from the start since we only had 2 nodes and that would create 125 pods on each worker node. According to the Kubernetes documentation the maximum recommended number of pods per node is 100.